export ESM code

dynst opened the pull request:
`"module"` is a non-standard field that Node.js doesn't read. It uses `"exports"`.

ESM code also needs the `.mjs` file extension.

https://nodejs.org/api/packages.html#determining-module-system

dynst committed:
add exports field to package.json

ESM projects can now import the tree-shakeable ESM code.
Node.js doesn't use the "module" field.

dynst committed:
keep redundant main field for webpack 4

This field is ignored by all modern tools when "exports" is present.

dynst committed:
use .cjs and .mjs file extensions

Daninet commented:
Do you know if this syntax is supported by FE tooling like Webpack 4/5, Rollup, esbuild, parcel etc?

dynst commented:
It's from Node v12 https://nodejs.org/en/blog/release/v12.7.0 so almost everything should support it by now.

Webpack 5 added support in 2020. https://webpack.js.org/blog/2020-10-10-webpack-5-release/ so version 4 didn't have it.

Rollup definitely has it. https://github.com/rollup/plugins/issues/208#issuecomment-1162367570 Since at least `node-resolve` v11. https://github.com/rollup/plugins/pull/540

dynst committed:
transpile 1 chunk per file for ESM build

index.ts -> index.mjs, argon2.ts -> argon2.mjs, rather than
bundled into one giant file.

dynst committed:
fix redundancy in tsconfig.json

es2017 is a superset of es7, which is a superset of es6.
So use es2017 consistently for target and lib.

dynst committed:
move type declarations out of lib/ subdirectory

now they live next to the files they describe,
thanks to rootDir.

tsconfig.json seems to be completely ignored and unused here.

bsidhom commented:
I haven't tested the above, but I can confirm that the current package cannot be implemented as a module by tools like tsdown. As a workaround, I've had to vendor the `./dist/index.esm.js` file to get it resolved at all.
